examples/verify_rule.cpp

```diff
@@ -0,0 +1,221 @@
+	return "off";
+}
+
+void log_cb(DDWAF_LOG_LEVEL level,
+			const char* function, const char* file, unsigned line,
+			const char* message, uint64_t)
+{
+	printf("[%s][%s:%s:%u]: %s\n", level_to_str(level), file, function, line, message);
+}
+
+#endif
+
+namespace YAML
+{
+	ddwaf_object node_to_arg(const Node& node)
+	{
+		switch (node.Type())
+		{
+			case NodeType::Sequence:
+			{
+				ddwaf_object arg;
+				ddwaf_object_array(&arg);
+				for (auto it = node.begin(); it != node.end(); ++it)
+				{
+					ddwaf_object child = node_to_arg(*it);
+					ddwaf_object_array_add(&arg, &child);
+				}
+				return arg;
+			}
+			case NodeType::Map:
+			{
+				ddwaf_object arg;
+				ddwaf_object_map(&arg);
+				for (auto it = node.begin(); it != node.end(); ++it)
+				{
+					std::string key    = it->first.as<std::string>();
+					ddwaf_object child = node_to_arg(it->second);
+					ddwaf_object_map_addl(&arg, key.c_str(), key.size(), &child);
+				}
+				return arg;
+			}
+			case NodeType::Scalar:
+			{
+				const std::string& value = node.Scalar();
+				ddwaf_object arg;
+				ddwaf_object_stringl(&arg, value.c_str(), value.size());
+				return arg;
+			}
+			case NodeType::Null:
+			case NodeType::Undefined:
+				ddwaf_object arg;
+				ddwaf_object_invalid(&arg);
+				return arg;
+		}
+		
+		throw std::runtime_error("Invalid YAML node type");
+	}
+	
+	template <>
+	struct as_if<ddwaf_object, void>
+	{
+		explicit as_if(const Node& node_) : node(node_) {}
+		ddwaf_object operator()() const { return node_to_arg(node); }
+		const Node& node;
+	};
+	
+}
+
+std::string read_rule_file(const std::string_view& filename)
+{
+	std::ifstream rule_file(filename.data(), std::ios::in);
+	if (!rule_file)
+	{
+		throw std::system_error(errno, std::generic_category());
+	}
+	
+	// Create a buffer equal to the file size
+	std::string buffer;
+	rule_file.seekg(0, std::ios::end);
+	buffer.resize(rule_file.tellg());
+	rule_file.seekg(0, std::ios::beg);
+	
+	rule_file.read(&buffer[0], buffer.size());
+	rule_file.close();
+	return buffer;
+}
+
+ddwaf_object convertRuleToRuleset(YAML::Node rulePayload)
+{
+	ddwaf_object rule   = rulePayload.as<ddwaf_object>(), root, version, array;
+	
+	ddwaf_object_map(&root);
+	ddwaf_object_array(&array);
+	ddwaf_object_array_add(&array, &rule);
+	
+	ddwaf_object_map_add(&root, "version", ddwaf_object_string(&version, "2.1"));
+	ddwaf_object_map_add(&root, "rules", &array);
+	return root;
+}
+
+bool runVectors(YAML::Node rule, ddwaf_handle handle, bool runPositiveMatches)
+{
+	bool success = true;
+	std::string ruleID = rule["id"].as<std::string>();
+	YAML::Node matches = rule["test_vectors"][runPositiveMatches ? "matches" : "no_matches"];
+	if (matches != nullptr)
+	{
+		size_t counter = 0;
+		for (YAML::const_iterator vector = matches.begin(); vector != matches.end(); ++vector, ++counter) {
+			ddwaf_object root = vector->as<ddwaf_object>();
+			if(root.type != DDWAF_OBJ_INVALID) {
+				ddwaf_context ctx = ddwaf_context_init(handle, NULL);
```

Comment:
You can pass `ddwaf_object_free` as the second parameter and avoid the call to `ddwaf_object_free`.